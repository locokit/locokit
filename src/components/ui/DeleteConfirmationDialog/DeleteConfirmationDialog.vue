<template>
  <lck-dialog
    :visible="visible"
    :header="$t('form.delete')"
    :isActionForm="true"
    :submitting="submitting"
    @input="$emit('input', value)"
    @close="$emit('close')"
  >
    <p>
      <template v-if="value && value.text">
        {{ $t('form.specificDeleteConfirmation') }} <em>{{ itemCategory }} {{ value.text }}</em>.
      </template>
      <template v-else>{{$t('form.deleteConfirmation')}}</template>
    </p>
  </lck-dialog>
</template>

<script>
import Dialog from '@/components/ui/Dialog/Dialog.vue'

export default {
  name: 'DeleteConfirmationDialog',
  components: {
    'lck-dialog': Dialog
  },
  props: {
    itemCategory: {
      type: String,
      default: ''
    },
    submitting: {
      type: Boolean,
      default: false
    },
    visible: {
      type: Boolean,
      default: false
    },
    value: {
      type: Object
    }
  }
}
</script>
