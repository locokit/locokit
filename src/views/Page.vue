<template>
  <div class="mx-2">
    <header class="text-2xl text-blue-600 font-medium my-4">
      {{ page && page.text }}
    </header>
    <div v-if="page && page.containers.length > 0">
      <div v-for="container in page.containers" :key="container.id">
        <Container :container="container"></Container>
      </div>
    </div>
  </div>
</template>

<script>
import {
  retrievePageWithContainersAndBlocks,
  retrieveViewDefinition,
  retrieveViewData
} from '@/store/visualize'
import Container from '@/components/visualize/Container/Container'

export default {
  name: 'Page',
  components: { Container },
  props: {
    pageId: {
      type: [String, Number], // param is string because its form url params
      required: true
    }
  },
  data () {
    return {
      page: null
    }
  },
  watch: {
    pageId: {
      immediate: true,
      async handler () {
        this.page = await retrievePageWithContainersAndBlocks(this.pageId)
      }
    },
    page (newVal) {
      // retrieve for each blocks the definition / data of the block
      if (!newVal || !newVal.containers || !newVal.containers.length > 0) return
      newVal.containers.forEach(container => {
        container.blocks.forEach(async block => {
          switch (block.type) {
            case 'TableView':
              this.$set(block, 'loading', true)
              this.$set(block, 'definition', await retrieveViewDefinition(block.settings?.id))
              this.$set(block, 'data', await retrieveViewData(block.definition.id))
              block.loading = false
              return
            default:
              console.log(block)
          }
        })
      })
    }
  }
}
</script>

<style scoped>
.title-page {
  font-size: 1.4rem;
  padding: 1em;
  font-weight: 600;
  color: #645f5f;
}
</style>
