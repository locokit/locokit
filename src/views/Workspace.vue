<template>
  <div class="p-grid h-full" v-if="workspaceContent">
    <div class="sidebar-menu-container p-col p-md-6 p-lg-2 lck-bg-primary o-auto h-max-full">
      <lck-chapters :chapters="workspaceContent.chapters" />
    </div>
    <div class="h-full p-col o-auto h-max-full">
      <router-view />
    </div>
  </div>
</template>

<script>
import { retrieveWorkspaceWithChaptersAndPages } from '@/store/visualize'
import Chapters from '@/components/visualize/Chapters/Chapters'
import { ROUTES_PATH } from '@/router/paths'

export default {
  name: 'Workspace',
  components: {
    'lck-chapters': Chapters
  },
  props: ['workspaceId'],
  data () {
    return {
      workspaceContent: []
    }
  },
  async mounted () {
    this.workspaceContent = await retrieveWorkspaceWithChaptersAndPages(this.workspaceId)
    if (!this.$route.path.includes('page') && this.workspaceContent.chapters.length > 0 && this.workspaceContent.chapters[0].pages.length > 0) {
      await this.$router.replace(`${ROUTES_PATH.WORKSPACE}/${this.workspaceId}${ROUTES_PATH.VISUALIZATION}/page/${this.workspaceContent.chapters[0].pages[0].id}`)
    }
  }
}
</script>
