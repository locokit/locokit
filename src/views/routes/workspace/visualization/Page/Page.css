
.lck-page-content {
  min-width: 20rem;
  transition-duration: 0.3s;
}

.editable-container {
  margin-bottom: 1rem;
  border: 2px solid var(--surface-lck-2);
  background-color: #ffffff;
  border-radius: var(--border-radius);
  box-shadow: 0px 0px 6px 0px rgba(194, 194, 194, 0.7);
  overflow: hidden;
}

.edit-container-line {
  padding-left: 0.5rem;
  display: flex;
  justify-content: space-between;
  border-bottom: 1px solid var(--primary-color);
}

.edit-container-line .p-button {
  color: var(--primary-color);
  height: 100%;
}

.edit-container-line .p-buttonset {
  flex-shrink: 0;
}

.edit-container-line .handle {
  cursor: move;
}

.p-button.new-block-button {
  color: var(--primary-color-darken);
  width: 100%;
  height: 3rem;
}

.p-button.new-container-button {
  color: var(--primary-color);
  height: 3rem;
  width: 100%;
  margin-bottom: 0.5rem;
  border: 1px solid var(--primary-color) !important;
}

::v-deep .editable-block .block-content {
  padding: 0.5rem;
}

::v-deep .edit-block-line {
  padding-left: 0.5rem;
}

.lck-container {
  border-radius: var(--border-radius);
}

.lck-container:target {
  scroll-margin-top: 50px;
}

.lck-container.editable-container .edit-container-line {
  flex-direction: row;
}

.lck-container.editable-container .edit-container-line .lck-color-title {
  color: var(--primary-color)
}

::v-deep .p-breadcrumb {
  background: unset;
  border: unset;
  padding-left: 0;
}
/* classic content */

.lck-layout-classic .lck-container {
  display: flex;
  flex-direction: column;
}

/* Contenu Centr√© */

.lck-layout-centered .lck-container {
  display: flex;
  flex-direction: column;
  max-width: 800px;
  margin: 0 auto;
  justify-content: space-between;
  overflow: auto;
  padding: 1rem;
}

.lck-layout-centered .lck-block {
  display: flex;
  flex: 0 1 100%;
}

.lck-layout-centered .lck-block.lck-media {
  justify-content: center;
}

/* Contenu Flex (2/n colonnes) */

.lck-layout-flex .lck-container div {
  display: flex;
  flex-direction: row;
  flex-grow: 1;
  flex-basis: 0;
  flex-wrap: wrap;
  column-gap: 1rem;
}

.lck-layout-flex .lck-container .lck-block.lck-media {
  justify-content: center;
}

.lck-layout-flex .lck-container .edit-container-line {
  align-self: flex-start;
  width: 100%;
}

@media (max-width: 900px) {
  .lck-layout-flex .lck-container div {
    flex-direction: column;
    flex-wrap: unset;
  }
}

/* Contenu Full */

.lck-layout-full {
  width: 100%;
  height: 100%;
}

.lck-layout-full .lck-container div {
  height: 100%;
  width: 100%;
  overflow: scroll;
}
