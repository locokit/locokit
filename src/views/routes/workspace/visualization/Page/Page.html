<div
    v-if="page"
    class="p-mx-2"
    :class="layoutPage"
  >
    <div
      class="lck-page-content"
      :style="{ marginRight: showUpdateSidebar ? editableSidebarWidth : 0 }"
    >
      <div
        v-if="page.hidden"
        class="lck-color-primary p-mb-4"
      >
        <p-breadcrumb
          :home="{ icon: 'pi pi-home', to: '/' }"
          :model="breadcrumb"
        />
      </div>

      <div class="lck-color-primary p-my-4">
        <h1>{{ page.text }}</h1>
      </div>

      <lck-nav-anchor-link
        v-if="isNavBarAnchorLinkDisplayed || editMode"
        :containers="page.containers"
        :editMode="editMode"
        @edit-nav="onNavAnchorLinkEditClick"
      />

      <draggable
        :key="page.id"
        v-model="page.containers"
        handle=".handle"
        @change="onContainerReorderClick"
      >
        <div
          v-for="container in page.containers"
          :id="container.id"
          :key="container.id"
          class="lck-container"
          :class="{
            'editable-container': editMode,
            'lck-elevation': container.elevation
          }"
        >
          <h2 v-if="container.display_title && !editMode" class="lck-color-title">{{ container.text }}</h2>
          <div v-if="editMode" class="edit-container-line">
            <h2 class="lck-color-title">{{ container.text }}</h2>
            <span class="p-buttonset">
              <p-button
                :title="$t('pages.workspace.container.drag')"
                class="p-button-lg p-button-text handle "
                icon="pi pi-ellipsis-v"
              />
              <p-button
                :title="$t('pages.workspace.container.edit')"
                class="p-button-lg p-button-text edit-container-button"
                icon="pi pi-pencil"
                @click="onContainerEditClick(container)"
              />
              <p-button
                :title="$t('pages.workspace.container.delete')"
                class="p-button-lg p-button-text remove-container-button"
                icon="pi pi-trash"
                @click="onContainerDeleteClick(container)"
              />
            </span>
          </div>
          <draggable
            :key="container.id"
            v-model="container.blocks"
            @change="onBlockReorderClick(container, $event)"
            handle=".handle-block"
          >
            <template v-for="block in container.blocks">
              <Block
                :key="block.id"
                v-if="editMode || isBlockDisplayed(block)"
                class="lck-block"
                :class="{
                  'p-mb-4': !editMode,
                }"
                :block="block"
                :content="getBlockContent(block)"
                :definition="getBlockDefinition(block)"
                :workspaceId="workspaceId"
                :autocompleteSuggestions="autocompleteSuggestions"
                :exporting="exporting"
                :cellState="cellState"
                :editMode="editMode"
                v-on="$listeners"

                @row-delete="onRowDelete(block, $event)"
                @row-duplicate="onRowDuplicate(block, $event)"

                @update-row="onUpdateCell(block, $event)"
                @update-cell="onUpdateCell(block, $event)"
                @update-content="onUpdateContentBlockTableView(block, $event)"
                @update-suggestions="onUpdateSuggestions"
                @sort="onSort(block, $event)"
                @open-detail="onPageDetail(block, $event)"
                @create-row="onCreateRow(block, $event)"
                @export-view-csv="onExportViewCSV(block)"
                @export-view-xls="onExportViewXLS(block)"
                @update-filters="onUpdateFilters(block, $event)"
                @update-block="onBlockEditClick(container, block)"
                @delete-block="onBlockDeleteClick(container, block)"

                @download-attachment="onDownloadAttachment"
                @upload-files="onUploadFiles(block, $event)"
                @remove-attachment="onRemoveAttachment(block, $event)"

                @go-to-page-detail="goToPage"
                @create-process-run="onTriggerProcess(block, $event)"

                @update-features="onGeoDataEdit(block, $event)"
                @remove-features="onGeoDataRemove(block, $event)"
              />
            </template>
          </draggable>
          <p-button
            v-if="editMode"
            :title="$t('pages.workspace.block.create')"
            icon="pi pi-plus"
            class="new-block-button p-button-text"
            @click="onBlockEditClick(container, { id: 'temp' })"
          />
        </div>
      </draggable>
      <p-button
        v-if="editMode"
        :title="$t('pages.workspace.container.create')"
        icon="pi pi-plus"
        class="new-container-button p-button-text"
        @click="onContainerEditClick({ id: 'temp' })"
      />
    </div>
    <update-sidebar
      :submitting="submitting"
      :showSidebar="showUpdateSidebar"
      :container="currentContainerToEdit"
      :block="currentBlockToEdit"
      :page="page"
      :width="editableSidebarWidth"
      :autocompleteSuggestions="editableAutocompleteSuggestions"
      :blockDisplayTableViewSuggestions="blockDisplayTableViewSuggestions"
      :blockDisplayFieldSuggestions="blockDisplayFieldSuggestions"
      :relatedChapterPages="relatedChapterPages"
      @update-container="onContainerEditInput"
      @update-block="onBlockEditInput"

      @add-new-block="onBlockEditClickFromSidebar"
      @edit-block="onBlockEditClickFromSidebar"
      @delete-block="onBlockDeleteClick(currentContainerToEdit, $event)"

      @add-new-container="onContainerEditClickFromSidebar"
      @edit-container="onContainerEditClickFromSidebar"
      @delete-container="onContainerDeleteClick($event)"

      @reset-current-block="onBlockEditClickFromSidebar"
      @reset-current-container="onContainerEditClickFromSidebar"
      @close="onCloseUpdateContainerSidebar"
      @search-table-view="onSearchTableView"

      @search-block-display-table-view="onSearchBlockDisplayTableView"
      @search-block-display-field="onSearchBlockDisplayField"
    />
    <delete-confirmation-dialog
      :submitting="submitting"
      :visible="dialogVisibility.containerDelete"
      :value="currentContainerToDelete"
      :itemCategory="$t('pages.workspace.container.title')"
      @close="onContainerDeleteClose"
      @input="onContainerDeleteInput"
    />
    <delete-confirmation-dialog
      :submitting="submitting"
      :visible="dialogVisibility.blockDelete"
      :value="currentBlockToDelete"
      :itemCategory="$t('pages.workspace.block.title')"
      fieldToDisplay="title"
      @close="onBlockDeleteClose"
      @input="onBlockDeleteInput"
    />
  </div>
