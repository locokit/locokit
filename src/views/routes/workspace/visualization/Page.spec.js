/* eslint-disable @typescript-eslint/camelcase */

import { shallowMount, createLocalVue } from '@vue/test-utils'
import { BLOCK_TYPE } from '@locokit/lck-glossary'
import VueRouter from 'vue-router'
import draggable from 'vuedraggable'

import { ROUTES_PATH, ROUTES_NAMES } from '@/router/paths'
import { lckServices } from '@/services/lck-api'

import Page from './Page.vue'
import Workspace from '@/views/routes/workspace/visualization/Workspace.vue'

import UpdateContainerSidebar from '@/components/visualize/UpdateContainerSidebar/UpdateContainerSidebar.vue'
import DeleteConfirmationDialog from '@/components/ui/DeleteConfirmationDialog/DeleteConfirmationDialog.vue'

// Mock external libraries

jest.mock('@locokit/lck-glossary', () => ({
  BLOCK_TYPE: {
    PARAGRAPH: 'Paragraph',
    TABLE_VIEW: 'TableView'
  },
  COLUMN_TYPE: {
  }
}))

jest.mock('date-fns')
jest.mock('file-saver')
jest.mock('vuedraggable')

// Mock primevue component
jest.mock('primevue/button', () => ({
  name: 'p-button',
  render: h => h('p-button')
}))

jest.mock('primevue/breadcrumb', () => ({
  name: 'p-breadcrumb',
  render: h => h('p-breadcrumb')
}))

// Mock error
class MockError extends Error {
  constructor (code, ...args) {
    super(args)
    this.code = code
  }
}

// Mock variables
const mockPages = {
  1: {
    id: '1',
    text: 'Page 1',
    createdAt: '2020-11-02T16:11:03.109Z',
    updatedAt: '2020-11-02T16:11:03.109Z',
    chapter_id: 'C1',
    position: 1,
    hidden: false,
    containers: [
      {
        id: '12',
        text: 'My second container',
        createdAt: '2020-11-02T16:11:03.109Z',
        updatedAt: '2020-11-02T16:11:03.109Z',
        page_id: '1',
        settings: null,
        position: 2,
        blocks: [
          {
            id: '121',
            createdAt: '2020-11-02T16:11:03.109Z',
            updatedAt: '2020-11-02T16:11:03.109Z',
            settings: {
              content: 'My paragraph content'
            },
            title: 'My paragraph block',
            type: BLOCK_TYPE.PARAGRAPH,
            position: 2,
            container_id: '11'
          },
          {
            id: '122',
            createdAt: '2020-11-02T16:11:03.109Z',
            updatedAt: '2020-11-02T16:11:03.109Z',
            settings: {
              id: 'ref_view'
            },
            title: 'My paragraph block',
            type: BLOCK_TYPE.TABLE_VIEW,
            position: 1,
            container_id: '11'
          }
        ]
      },
      {
        id: '11',
        text: 'My first container',
        createdAt: '2020-11-02T16:11:03.109Z',
        updatedAt: '2020-11-02T16:11:03.109Z',
        page_id: '1',
        settings: null,
        position: 1,
        blocks: []
      }
    ]
  },
  2: {
    id: '2',
    text: 'Page 2',
    createdAt: '2020-11-02T16:11:03.109Z',
    updatedAt: '2020-11-02T16:11:03.109Z',
    chapter_id: 'C2',
    position: 2,
    hidden: true,
    containers: []
  }
}

function mockDeepCloneObject (object) {
  return JSON.parse(JSON.stringify(object))
}

// Mock lck functions

jest.mock('@/services/lck-api', () => ({
  lckServices: {
    container: {
      patch: jest.fn((id, data) =>
        ({ ...mockPages['1'].containers.find(container => container.id === id), ...data })),
      create: jest.fn(data =>
        ({ ...mockPages['1'].containers[0], ...data, id: '22' })),
      remove: jest.fn()
    }
  },
  lckHelpers: {
    searchItems: jest.fn(() => ([
      { label: 'A', value: 1 },
      { label: 'B', value: 2 },
      { label: 'C', value: 3 }
    ])),
    exportTableRowData: jest.fn(() => 'CSV_EXPORT')
  }
}))

jest.mock('@/store/visualize', () => ({
  retrievePageWithContainersAndBlocks: jest.fn((pageId) => mockDeepCloneObject(mockPages[pageId])),
  retrieveViewDefinition: jest.fn(() => ({ id: 'definitionId' })),
  retrieveViewData: jest.fn(() => ('content'))
}))

jest.mock('@/store/database')

jest.mock('@/router/paths', () => ({
  ROUTES_PATH: {
    WORKSPACE: '/workspace',
    VISUALIZATION: '/visualization'
  },
  ROUTES_NAMES: {
    VISUALIZATION: 'WorkspaceVisualization',
    PAGE: 'Page'
  }
}))

jest.mock('@/views/routes/workspace/visualization/Workspace.vue', () => ({
  name: 'Workspace',
  render: h => h('workspace'),
  computed: {
    sidebarItems: () => []
  }
}))

jest.mock('@/components/visualize/Block/Block', () => ({
  name: 'Block',
  render: h => h('block')
}))

// Mock routes
const mockRoutes = [
  {
    path: ROUTES_PATH.WORKSPACE + '/:workspaceId' + ROUTES_PATH.VISUALIZATION,
    name: ROUTES_NAMES.VISUALIZATION,
    component: Page,
    props: true,
    children: [{
      name: ROUTES_NAMES.PAGE,
      path: 'page/:pageId',
      props: true,
      component: Page
    }]
  }
]

// Tests

describe('Page', () => {
  // For vue router
  const localVue = createLocalVue()
  localVue.use(VueRouter)
  const router = new VueRouter({ routes: mockRoutes })

  // Default workspace component configuration
  function globalComponentParams (pageId = '1') {
    return {
      localVue,
      router,
      propsData: { pageId },
      mocks: {
        t: key => key,
        $t: key => key,
        $toast: {
          add: jest.fn()
        }
      },
      parentComponent: Workspace
    }
  }

  describe('Watchers', () => {
    describe('Load the specified page with containers', () => {
      let wrapper

      beforeAll(async () => {
        wrapper = await shallowMount(Page, globalComponentParams())
      })
      it('Get the right id', () => {
        expect(wrapper.vm.page.id).toBe(mockPages['1'].id)
      })
      it('Get the right text property', () => {
        expect(wrapper.vm.page.text).toBe(mockPages['1'].text)
      })
      it('Get the right chapter_id property', () => {
        expect(wrapper.vm.page.chapter_id).toBe(mockPages['1'].chapter_id)
      })
      it('Get the right position property', () => {
        expect(wrapper.vm.page.position).toBe(mockPages['1'].position)
      })
      it('Get the right hidden property', () => {
        expect(wrapper.vm.page.hidden).toBe(mockPages['1'].hidden)
      })
      describe('Get the right containers', () => {
        it('The containers are sorted', () => {
          expect(wrapper.vm.page.containers[0].id).toBe(mockPages['1'].containers[1].id)
          expect(wrapper.vm.page.containers[1].id).toBe(mockPages['1'].containers[0].id)
        })
        it('Check the empty container (without block)', () => {
          expect(wrapper.vm.page.containers[0]).toStrictEqual(mockPages['1'].containers[1])
        })
        it('The blocks inside each container are sorted', () => {
          expect(wrapper.vm.page.containers[1].blocks[0].id).toBe(mockPages['1'].containers[0].blocks[1].id)
          expect(wrapper.vm.page.containers[1].blocks[1].id).toBe(mockPages['1'].containers[0].blocks[0].id)
        })
        it('Get the right properties for the Paragraph block (default)', () => {
          expect(wrapper.vm.page.containers[1].blocks[1]).toMatchObject(mockPages['1'].containers[0].blocks[0])
        })
        it('Get the right properties for the TableView block', () => {
          expect(wrapper.vm.page.containers[1].blocks[0]).toMatchObject(mockPages['1'].containers[0].blocks[1])
          expect(wrapper.vm.page.containers[1].blocks[0].definition).toStrictEqual({ id: 'definitionId' })
          expect(wrapper.vm.page.containers[1].blocks[0].content).toStrictEqual('content')
        })
      })
    })
    describe('Load the specified page without any containers', () => {
      it('Get the right containers', async () => {
        const wrapper = await shallowMount(Page, globalComponentParams('2'))
        expect(wrapper.vm.page.containers).toStrictEqual([])
      })
    })
  })

  describe('Container CRUD', () => {
    describe('Display edit options', () => {
      let wrapper

      beforeAll(async () => {
        wrapper = await shallowMount(Page, globalComponentParams())
        wrapper.setProps({ editMode: true })
      })

      it('Display the button to add a container', () => {
        expect(wrapper.get('.new-container-button').exists()).toBe(true)
      })

      it('Display the containers names', () => {
        const containersNames = wrapper.findAll('.edit-container-line h2')
        expect(containersNames.at(0).text()).toBe(mockPages['1'].containers[1].text)
        expect(containersNames.at(1).text()).toBe(mockPages['1'].containers[0].text)
      })

      it('Display the drag buttons', () => {
        expect(wrapper.findAll('.edit-container-line .handle').length).toBe(2)
      })

      it('Display the edit buttons', () => {
        expect(wrapper.findAll('.edit-container-line .edit-button').length).toBe(2)
      })

      it('Display the remove buttons', () => {
        expect(wrapper.findAll('.edit-container-line .remove-button').length).toBe(2)
      })
    })

    describe('Events', () => {
      let wrapper
      let containerSidebarWrapper
      let deleteConfirmationWrapper

      const newContainerName = 'newContainerName'
      const firstDisplayedContainer = mockPages['1'].containers[1]
      const secondDisplayedContainer = mockPages['1'].containers[0]

      beforeEach(async () => {
        wrapper = await shallowMount(Page, { ...globalComponentParams(), propsData: { pageId: '1', editMode: true } })
        containerSidebarWrapper = wrapper.findComponent(UpdateContainerSidebar)
        deleteConfirmationWrapper = wrapper.findComponent(DeleteConfirmationDialog)
      })

      describe('Add a new container', () => {
        beforeEach(() => {
          lckServices.container.create.mockClear()
        })

        it('Display the update container sidebar with an empty object when the new container button is clicked', async () => {
          await wrapper.find('.new-container-button').vm.$emit('click')
          expect(containerSidebarWrapper.props('container')).toStrictEqual({})
          expect(containerSidebarWrapper.props('showSidebar')).toBe(true)
        })

        it('Create a new container if the update-container event is emitted from the update container sidebar with a new container', async () => {
          await wrapper.find('.new-container-button').vm.$emit('click')
          await containerSidebarWrapper.vm.$emit('update-container', { text: newContainerName })
          // Send API request
          expect(lckServices.container.create).toHaveBeenCalledWith({ text: newContainerName, page_id: mockPages['1'].id })
          // Update the component data
          const newContainer = wrapper.vm.page.containers[2]
          expect(newContainer).toBeDefined()
          expect(newContainer.text).toBe(newContainerName)
        })
      })

      describe('Update an existing container', () => {
        beforeEach(() => {
          lckServices.container.patch.mockClear()
        })

        it('Display the update container sidebar with the specified container when the edit button is clicked', async () => {
          await wrapper.find('.edit-button').vm.$emit('click', firstDisplayedContainer)
          expect(containerSidebarWrapper.props('container')).toStrictEqual(firstDisplayedContainer)
          expect(containerSidebarWrapper.props('showSidebar')).toBe(true)
        })

        it('Hide the update container sidebar if it emits the close event', async () => {
          // Display the sidebar
          await containerSidebarWrapper.setProps({ showSidebar: true })
          // Hide it
          await containerSidebarWrapper.vm.$emit('close')
          expect(containerSidebarWrapper.props('showSidebar')).toBe(false)
        })

        it('Update the container if the update-container is emitted with an existing container from the sidebar', async () => {
          await wrapper.find('.edit-button').vm.$emit('click', firstDisplayedContainer)
          await containerSidebarWrapper.vm.$emit('update-container', { id: firstDisplayedContainer.id, text: newContainerName })
          // Send API request
          expect(lckServices.container.patch).toHaveBeenCalledWith(firstDisplayedContainer.id, { text: newContainerName })
          // Update the component data
          expect(
            wrapper.vm.page.containers.find(
              container => container.id === firstDisplayedContainer.id &&
              container.text === newContainerName)
          ).toBeDefined()
        })

        it('Display a toast if an error is occured', async () => {
          const spyOnToast = jest.spyOn(wrapper.vm, 'displayToastOnError')
          lckServices.container.patch.mockImplementationOnce(() => { throw new Error() })
          await wrapper.find('.edit-button').vm.$emit('click', firstDisplayedContainer)
          await containerSidebarWrapper.vm.$emit('update-container', { id: firstDisplayedContainer.id, text: newContainerName })
          expect(spyOnToast).toHaveBeenCalledTimes(1)
        })
      })

      describe('Delete a container', () => {
        beforeEach(() => {
          lckServices.container.remove.mockClear()
        })

        it('Display the confirmation dialog with the specified container when the remove button is clicked', async () => {
          await wrapper.find('.remove-button').vm.$emit('click')
          expect(deleteConfirmationWrapper.props('visible')).toBe(true)
          expect(deleteConfirmationWrapper.props('value')).toStrictEqual(firstDisplayedContainer)
        })

        it('Hide the confirmation dialog if the close event is emitted', async () => {
          // Display the dialog
          await deleteConfirmationWrapper.setProps({ visible: true })
          // Hide it
          await deleteConfirmationWrapper.vm.$emit('close')
          expect(deleteConfirmationWrapper.props('visible')).toBe(false)
        })

        it('Delete a container if the input event is emitted with an existing container', async () => {
          await wrapper.find('.remove-button').vm.$emit('click')
          await deleteConfirmationWrapper.vm.$emit('input', firstDisplayedContainer)
          // Send API request
          expect(lckServices.container.remove).toHaveBeenCalledWith(firstDisplayedContainer.id)
          // Update the component data
          expect(
            wrapper.vm.page.containers.find(container => container.id === firstDisplayedContainer.id)
          ).toBeUndefined()
        })

        it('Do nothing if the input event is emitted with an empty container', async () => {
          await wrapper.find('.remove-button').vm.$emit('click')
          await deleteConfirmationWrapper.vm.$emit('input', {})
          expect(
            wrapper.vm.page.containers.find(container => container.id === firstDisplayedContainer.id)
          ).toBeDefined()
        })

        it('Do nothing if the input event is emitted with an undefined container', async () => {
          await wrapper.find('.remove-button').vm.$emit('click')
          await deleteConfirmationWrapper.vm.$emit('input')
          expect(
            wrapper.vm.page.containers.find(container => container.id === firstDisplayedContainer.id)
          ).toBeDefined()
        })

        it('Do nothing if the input event is emitted with an unknown container', async () => {
          await wrapper.find('.remove-button').vm.$emit('click')
          await deleteConfirmationWrapper.vm.$emit('input', { ...firstDisplayedContainer, id: '-1' })
          expect(
            wrapper.vm.page.containers.find(container => container.id === firstDisplayedContainer.id)
          ).toBeDefined()
        })

        it('Display a toast if an error is occured', async () => {
          const spyOnToast = jest.spyOn(wrapper.vm, 'displayToastOnError')
          lckServices.container.remove.mockImplementationOnce(() => { throw new Error() })
          await wrapper.find('.remove-button').vm.$emit('click')
          await deleteConfirmationWrapper.vm.$emit('input', { id: firstDisplayedContainer.id })
          expect(spyOnToast).toHaveBeenCalledTimes(1)
        })
      })

      describe('Reorder the containers', () => {
        beforeEach(() => {
          lckServices.container.patch.mockClear()
        })
        it('Save the new containers positions if the change event is emitted (oldIndex < newIndex)', async () => {
          await wrapper.findComponent(draggable).vm.$emit('change', { moved: { oldIndex: 0, newIndex: 1 } })
          expect(lckServices.container.patch).toHaveBeenCalledTimes(2)
          expect(lckServices.container.patch).toHaveBeenNthCalledWith(1, firstDisplayedContainer.id, { position: 0 })
          expect(lckServices.container.patch).toHaveBeenNthCalledWith(2, secondDisplayedContainer.id, { position: 1 })
        })
        it('Save the new containers positions if the change event is emitted (oldIndex > newIndex)', async () => {
          await wrapper.findComponent(draggable).vm.$emit('change', { moved: { oldIndex: 1, newIndex: 0 } })
          expect(lckServices.container.patch).toHaveBeenCalledTimes(2)
          expect(lckServices.container.patch).toHaveBeenNthCalledWith(1, firstDisplayedContainer.id, { position: 0 })
          expect(lckServices.container.patch).toHaveBeenNthCalledWith(2, secondDisplayedContainer.id, { position: 1 })
        })
        it('Do not save the new containers positions if the drag over event is not a move event ', async () => {
          await wrapper.findComponent(draggable).vm.$emit('change', { added: { newIndex: 0 } })
          expect(lckServices.container.patch).toHaveBeenCalledTimes(0)
        })
        it('Display a toast if an error is occured', async () => {
          const spyOnToast = jest.spyOn(wrapper.vm, 'displayToastOnError')
          lckServices.container.patch.mockRejectedValueOnce(new MockError())
          await wrapper.findComponent(draggable).vm.$emit('change', { moved: { oldIndex: 0, newIndex: 1 } })
          process.nextTick(() => {
            expect(spyOnToast).toHaveBeenCalledTimes(1)
          })
        })
      })
    })
  })

  describe('Methods', () => {
    describe('displayToastOnError', () => {
      let wrapper

      beforeAll(async () => {
        wrapper = await shallowMount(Page, globalComponentParams())
        wrapper.vm.$toast.add.mockClear()
      })

      afterEach(() => {
        wrapper.vm.$toast.add.mockClear()
      })

      it('Display a toast with the specified parameters for an error without code', () => {
        wrapper.vm.displayToastOnError('summary', new Error('an error without a code'))
        expect(wrapper.vm.$toast.add).toHaveBeenLastCalledWith(
          expect.objectContaining({
            summary: 'summary',
            detail: 'error.basic'
          })
        )
      })

      it('Display a toast with the specified parameters for an error with a code', () => {
        wrapper.vm.displayToastOnError('summary', new MockError(404, 'an error with a code'))
        expect(wrapper.vm.$toast.add).toHaveBeenLastCalledWith(
          expect.objectContaining({
            summary: 'summary',
            detail: 'error.http.404'
          })
        )
      })
    })
  })
})
