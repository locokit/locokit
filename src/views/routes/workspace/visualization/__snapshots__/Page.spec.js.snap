// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Create new records (blocks Set) match snapshot and display all blocks 1`] = `
<div class="p-mx-2 lck-layout-classic">
  <div class="lck-page-content" style="margin-right: 0px;">
    <!---->
    <div class="lck-color-primary p-my-4">
      <h1>Page 6</h1>
    </div>
    <!---->
    <draggable-stub value="[object Object]" clone="function(t){return t}" element="div" handle=".handle" class="lck-container-parent">
      <div id="16" class="lck-container">
        <!---->
        <!---->
        <draggable-stub value="[object Object],[object Object]" clone="function(t){return t}" element="div" handle=".handle-block" class="lck-block-parent">
          <block-stub block="[object Object]" workspaceid="toto" content="[object Object]" definition="[object Object]" cellstate="[object Object]" secondarysources="[object Object]" class="lck-block p-pb-4"></block-stub>
          <block-stub block="[object Object]" workspaceid="toto" cellstate="[object Object]" secondarysources="[object Object]" class="lck-block p-pb-4" definition="[object Object]" content="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"></block-stub>
        </draggable-stub>
        <!---->
      </div>
    </draggable-stub>
    <!---->
  </div>
  <update-sidebar-stub container="[object Object]" block="[object Object]" page="[object Object]" width="30rem" relatedchapterpages=""></update-sidebar-stub>
</div>
`;
